### PR TITLE
Use multihash for encoding the attestation user_data

### LINKED ISSUE #23: user_data hard to parse
I didn't catch this in review, but the way we're storing hashes in the `user_data` section of the attestation document is tricky to parse. Although we prepend a hash key so we can update the digest algorithm and use a semicolon as a field separator, the hash itself is binary. That means one can't split on the (semi)colon characters to separate each hash and its key. Instead one must prefix-match on the keys and then use known hash lengths to split the fields.

Parsing would be more convenient if we hex-encoded the hash digest values so the punctuation can be unambiguous. This would reduce the number of (256-bit) hashes we could include from 25 to 14, both of whom seem like plenty.

Less convenient, we'd need to support both schemes in brave-core across a transition period.

### PR DESCRIPTION
**Pull request by rillian**
Because the `hashSeparator` character can occur within the binary hash digest values representing the https cert and optional app key, it's not useful for dividing the user_data value into separate fields.

Instead use the multihash encoding, which prefixes two bytes. The first (0x12) define the hash digest as SHA-256 just as the previous "sha256:" prefix did. The second (0x20) marks the digest length. This more compactly delimits the values while allowing the same extensibility.

Resolves #23

### CODE REVIEW

**Comment by rillian** [comment]
NB we shouldn't deploy this to the star-randsrv app until after https://github.com/brave/brave-core/pull/18954 has made it into stable release.

File: attestation.go [hunk 1]
```diff
@@ -39,14 +38,13 @@ type AttestationHashes struct {
...
 
-// Serialize returns a byte slice that contains our concatenated hashes.  Note
-// that all hashes are always present.  If a hash was not initialized, it's set
-// to 0-bytes.
+// Serialize returns a byte slice that contains our concatenated hashes.
+// hashPrefix defines the hash type and length.  Note that all hashes are
+// always present.  If a hash was not initialized, it's set to 0-bytes.
 func (a *AttestationHashes) Serialize() []byte {
-	str := fmt.Sprintf("%s%s%s%s%s",
+	str := fmt.Sprintf("%s%s%s%s",
```
Comment (DJAndries) [inline attestation.go:45]: i'm unsure if `fmt.Sprintf` is a robust fit for creating a binary value. it may be better to use `append` or something similar

Comment (rillian) [reply attestation.go:45]: There's also a go-multihash package we could use to perform the encoding at insert time, so we only need to concatenate the two slices. That might be cleaner overall.

Comment (DJAndries) [reply attestation.go:45]: > There's also a [go-multihash](https://pkg.go.dev/github.com/multiformats/go-multihash) package we could use to perform the encoding at insert time

Yeah, i suppose you could, but I think what you wrote above looks fine & lightweight enough. You could probably drop the `[:]` and just keep the `...` at the end of the slice that is to be appended. A shortened version could look something like this:
```
serialized := []byte{}
serialized = append(serialized, append(hashPrefix, a.tlsKeyHash...)...)
serialized = append(serialized, append(hashPrefix, a.appKeyHash...)...)
return serialized
```

Comment (DJAndries) [reply attestation.go:45]: or even:

```
serialized := append(hashPrefix, a.tlsKeyHash...)
serialized = append(serialized, append(hashPrefix, a.appKeyHash...)...)
return serialized
```

Comment (rillian) [reply attestation.go:45]: The `[:]` is required because the AttestationHashes members are arrays, not slices.

Comment (DJAndries) [reply attestation.go]: good point, edited suggestion above

Comment (rillian) [reply attestation.go]: I actually like your first suggestion better, since it aligns the two calls and makes it more obvious what's different between the lines. See f04c24d.

**Review by NullHypothesis — APPROVED** [review]
Looks good.

**Comment by rillian** [comment]
The corresponding brave-core change is in 1.54 scheduled for stable release July 18. I'll hold off on merging until then to avoid blocking other work.

**Comment by DJAndries** [comment]
Could we please hold off for a few months? Even though 1.54 will be released in the coming weeks, it does take a while for the majority of clients to update (i.e. Android users)

**Comment by rillian** [comment]
Looking at data from last week, we're about halfway through the transition (300k supporting multihash vs 400k not), so we should revisit after the next release.

**Comment by rillian** [comment]
@DJAndries Are you ok with merging this now? The browser-side change has been deployed for ~7 months, and attestation has been disabled for about the same amount of time.

**Comment by DJAndries** [comment]
> @DJAndries Are you ok with merging this now? The browser-side change has been deployed for ~7 months, and attestation has been disabled for about the same amount of time.

i'm cool with it :+1:

**Comment by rillian** [comment]
> i'm cool with it 👍

Great. Could you mark it as approved please, so I can merge?
